### PR TITLE
draft per-token processor for huggingface_pipeline (stop sequences, streaming)

### PR DESCRIPTION
**Pull request by xloem**
I did a little work to provide for stop sequences and streaming with huggingface_pipeline.

I found the tests and discovered they didn't run presently for me; I made a small change to the constructors to provide for pipeline kwargs so that they would. It still outputs warnings which I think would be correctly resolved by handing parameters off to the pipeline call during generation. A different issue.

The first streaming interface I bumped into was the callback one, so I implemented this one. After later finding the generator interface, I do think it looks nicer, although I see it is marked as beta.

I still have a lot to learn about the architecture, and my code-fu may be faltering; however this may be useful enough to be helpful for the huggingface pipeline code.

### CODE REVIEW

**Comment by quantumsheep** [comment]
Hey! Is this planned to be merged?

**Comment by xloem** [comment]
I'm no longer using langchain but I observe this file has moved to libs/langchain/langchain/llms/huggingface_pipeline.py . Might make sense to rebase or reimplement, or remerge with moving the files to the new paths.

**Comment by poojitharamachandra** [comment]
hi, is this implemented in the latest version, if not, when are we going to get this?

also, I took the pr, made local changes , but I still don't get streaming onto the chainlit UI

**Comment by efriis** [comment]
Hey @xloem thanks for the note! I'll close this for now because it's not mergable in its current state, and it looks like there are already some issues referencing the PR if someone else ends up taking a stab at this :)

Let me know if you want to own those updates, and we can reopen this!

**Comment by xloem** [comment]
Basically, at the time of implementation (and this is likely still the case but I have not looked recently), the HuggingFace generation interface does not provide for streaming. However, it does provide for two callbacks that can be used to implement things like streaming: LogitsProcessor ([doc definition](https://huggingface.co/docs/transformers/v4.35.0/en/internal/generation_utils#transformers.LogitsProcessor) / [doc use](https://huggingface.co/docs/transformers/v4.35.0/en/main_classes/text_generation#transformers.GenerationMixin.generate.logits_processor) / [source definition](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py) / [source use](https://github.com/huggingface/transformers/blob/cc3e4781854a52cf090ffde28d884a527dab6708/src/transformers/generation/utils.py#L2627)) and StoppingCriteria ([doc definition](https://huggingface.co/docs/transformers/v4.35.0/en/internal/generation_utils#transformers.StoppingCriteria) / [doc use](https://huggingface.co/docs/transformers/v4.35.0/en/main_classes/text_generation#transformers.GenerationMixin.generate.stopping_criteria) / [source definition](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/stopping_criteria.py) / [source use](https://github.com/huggingface/transformers/blob/cc3e4781854a52cf090ffde28d884a527dab6708/src/transformers/generation/utils.py#L2675)). Either one or both of these can be passed to the generation function via the keyword parameters `logits_processor` and `stopping_criteria` to received callback on every token.

They differ in that `logits_processor` is passed the latest logprobs (and its intended function is to mutate them), whereas `stopping_criteria` is called with the latest token after selection from the logprobs (and its intended function is to determine if generation is complete).

It looks like in my PR here, what I did was make a [very simple member function that redecoded the logits](https://github.com/langchain-ai/langchain/pull/1222/files#diff-c79ffe136348560c76049fda7ba9d087df2ed0dce569336d3e427020c4056093R220) and handed the string off as it grew to a [more complex member function that performed streaming and stopping](https://github.com/langchain-ai/langchain/pull/1222/files#diff-c79ffe136348560c76049fda7ba9d087df2ed0dce569336d3e427020c4056093R226). Rather than implementing a `stopping_criteria`, I mutated the logits to cause termination of the stream. I may have been unaware of stopping criteria, or I may have found it simpler in some way for me to do this. I then [wrap the function in a list and pass it as `logits_processor` when the generation function is called](https://github.com/langchain-ai/langchain/pull/1222/files#diff-c79ffe136348560c76049fda7ba9d087df2ed0dce569336d3e427020c4056093R201) (here a pipeline object, which hands its parameters off).

**Comment by xloem** [comment]
Looks like there is indeed streaming functionality in transformers now, example at https://huggingface.co/docs/transformers/v4.35.0/en/internal/generation_utils#transformers.TextIteratorStreamer
